xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.rbac.engine;
+
+import com.google.api.expr.v1alpha1.Expr;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.Descriptors.Descriptor;
+import io.envoyproxy.envoy.config.rbac.v2.Policy;
+import io.envoyproxy.envoy.config.rbac.v2.RBAC;
+import io.envoyproxy.envoy.config.rbac.v2.RBAC.Action;
+import io.grpc.xds.internal.rbac.engine.cel.Activation;
+import io.grpc.xds.internal.rbac.engine.cel.DefaultDispatcher;
+import io.grpc.xds.internal.rbac.engine.cel.DefaultInterpreter;
+import io.grpc.xds.internal.rbac.engine.cel.DescriptorMessageProvider;
+import io.grpc.xds.internal.rbac.engine.cel.Dispatcher;
+import io.grpc.xds.internal.rbac.engine.cel.IncompleteData;
+import io.grpc.xds.internal.rbac.engine.cel.Interpreter;
+import io.grpc.xds.internal.rbac.engine.cel.InterpreterException;
+import io.grpc.xds.internal.rbac.engine.cel.RuntimeTypeProvider;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * CEL Evaluation Engine is part of the authorization framework in gRPC.
+ * CEL Evaluation Engine takes one or two Envoy RBAC policies as input
+ * and uses CEL library to evaluate the condition field 
+ * inside each RBAC policy based on the provided Envoy Attributes. 
+ * CEL Evaluation Engine will generate an authorization decision which
+ * could be ALLOW, DENY or UNKNOWN.
+ * 
+ * <p>Use as in:
+ * 
+ * <pre>
+ *  AuthorizationEngine engine = new AuthorizationEngine(rbacPolicies);
+ *  AuthorizationDecision result = engine.evaluate();
+ * </pre>
+ */
+public class AuthorizationEngine<ReqT, RespT> {
+  private static final Logger log = Logger.getLogger(AuthorizationEngine.class.getName());
+  
+  /**
+   * RbacEngine is an inner class that holds RBAC action 
+   * and conditions of RBAC policy.
+   */
+  @SuppressWarnings("ClassCanBeStatic")
+  private class RbacEngine {
+    private final Action action;
+    private final ImmutableMap<String, Expr> conditions;
```

Comment:
Can we directly use RBAC.Conditions here? If so, we probably can avoid creating our own map and can just use whatever provided by RBAC.